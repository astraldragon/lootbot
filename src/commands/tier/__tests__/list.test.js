jest.mock("discord-akairo");

jest.mock("../../../models", () => ({
  Tier: { findAll: jest.fn("findAll") }
}));

const { Tier } = require("../../../models");
const { checkManagePermissions } = require("../../../support");
const { Command } = require("discord-akairo");
const ListCommand = require("../list");

describe("commands/tier/list", () => {
  beforeEach(() => {
    this.command = new ListCommand();

    this.msg = {
      guild: { id: "1234" },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should have the proper permission check", () => {
    expect(Command.mock.calls[0][1].options.permissions).toBe(
      checkManagePermissions
    );
  });

  it("should fetch the tiers for the guild", () => {
    Tier.findAll.mockReturnValue([]);
    this.command.exec(this.msg);

    expect(Tier.findAll).toHaveBeenCalledWith({
      where: { guild: this.msg.guild.id },
      order: [["weight", "DESC"]]
    });
  });

  it("should notify when no loot is found for the guild", async () => {
    Tier.findAll.mockReturnValue([]);
    await this.command.exec(this.msg);
    expect(this.msg.channel.send).toHaveBeenCalledWith("No tiers found");
  });

  it("should list the tiers", async () => {
    Tier.findAll.mockReturnValue([
      {
        name: "Common",
        color: 4737096,
        image: "test.png",
        weight: 10,
        luckyWeight: 20
      },
      {
        name: "Uncommon",
        color: 4171061,
        image: "test2.png",
        weight: 10,
        luckyWeight: 20
      }
    ]);

    await this.command.exec(this.msg);
    expect(this.msg.channel.send).toMatchSnapshot();
  });

  it("should notify if any errors occur", async () => {
    Tier.findAll.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      "An error occurred listing tiers"
    );
  });

  it("should properly format a color that would contain 0s in hex", async () => {
    Tier.findAll.mockReturnValue([
      {
        name: "Common",
        color: 17, // #000001
        image: "test.png",
        weight: 10,
        luckyWeight: 20
      }
    ]);

    await this.command.exec(this.msg);

    expect(this.msg.channel.send).toMatchSnapshot();
  });
});
