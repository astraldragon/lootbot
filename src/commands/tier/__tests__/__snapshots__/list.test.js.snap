// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/tier/list should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "tier-list",
      Object {
        "aliases": Array [
          "tier-list",
          "tl",
        ],
        "category": "Tier",
        "channelRestriction": "guild",
        "description": Object {
          "content": "List the glorious tiers",
          "examples": Array [
            "tier-list",
          ],
        },
        "options": Object {
          "permissions": [Function],
        },
      },
    ],
  ],
}
`;

exports[`commands/tier/list should list the tiers 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "
__**Common**__
Color: #484848
Image: <test.png>
Weight: 50.00% (10 in 20)
Lucky Weight: 50.00% (20 in 40)

__**Uncommon**__
Color: #3fa535
Image: <test2.png>
Weight: 50.00% (10 in 20)
Lucky Weight: 50.00% (20 in 40)
",
    ],
  ],
}
`;
